scheduler/plugin/host.go

```diff
@@ -26,79 +26,101 @@ import (
+		WithParameterNames("buf", "buf_len").Export(k8sSchedulerStatusReason).
+		Instantiate(ctx)
+}
+
+// paramsKey is a context.Context value associated with a params
 // pointer to the current request.
-type filterParamsKey struct{}
+type paramsKey struct{}
 
-type filterParams struct {
-	pod      *v1.Pod
+type params struct {
+	// pod is used by guest.filterFn and guest.scoreFn
+	pod *v1.Pod
+
+	// nodeInfo is used by guest.filterFn
 	nodeInfo *framework.NodeInfo
-	// reason is a field to avoid compiler-specific malloc/free functions, and
-	// to avoid having to deal with out-params because TinyGo only supports a
+
+	// nodeName is used by guest.scoreFn
+	nodeName string
+
+	// reason returned by all guest exports.
+	//
+	// It is a field to avoid compiler-specific malloc/free functions, and to
+	// avoid having to deal with out-params because TinyGo only supports a
 	// single result.
 	reason string
 }
 
-func filterParamsFromContext(ctx context.Context) *filterParams {
-	return ctx.Value(filterParamsKey{}).(*filterParams)
+func paramsFromContext(ctx context.Context) *params {
+	return ctx.Value(paramsKey{}).(*params)
 }
 
-// k8sSchedulerStatusReasonFn is a function used by the wasm guest to set the
-// framework.Status reason.
-func k8sSchedulerStatusReasonFn(ctx context.Context, mod wazeroapi.Module, stack []uint64) {
-	ptr := uint32(stack[0])
-	size := bufLimit(stack[1])
+func k8sApiNodeInfoNodeFn(ctx context.Context, mod wazeroapi.Module, stack []uint64) {
+	buf := uint32(stack[0])
+	bufLimit := bufLimit(stack[1])
 
-	var reason string
-	if b, ok := mod.Memory().Read(ptr, size); !ok {
-		// don't panic if we can't read the message.
-		reason = "BUG: out of memory reading message"
-	} else {
-		reason = string(b)
-	}
-	filterParamsFromContext(ctx).reason = reason
+	node := paramsFromContext(ctx).nodeInfo.Node()
+
+	stack[0] = uint64(marshalIfUnderLimit(mod.Memory(), node, buf, bufLimit))
 }
 
-func k8sApiNodeInfoNodeFn(ctx context.Context, mod wazeroapi.Module, stack []uint64) {
+func k8sApiNodeNameFn(ctx context.Context, mod wazeroapi.Module, stack []uint64) {
 	buf := uint32(stack[0])
 	bufLimit := bufLimit(stack[1])
 
-	node := filterParamsFromContext(ctx).nodeInfo.Node()
+	nodeName := paramsFromContext(ctx).nodeName
 
-	stack[0] = uint64(marshalIfUnderLimit(mod.Memory(), node, buf, bufLimit))
+	stack[0] = uint64(writeStringIfUnderLimit(mod.Memory(), nodeName, buf, bufLimit))
 }
 
 func k8sApiPodFn(ctx context.Context, mod wazeroapi.Module, stack []uint64) {
 	buf := uint32(stack[0])
 	bufLimit := bufLimit(stack[1])
 
-	pod := filterParamsFromContext(ctx).pod
+	pod := paramsFromContext(ctx).pod
 	stack[0] = uint64(marshalIfUnderLimit(mod.Memory(), pod, buf, bufLimit))
 }
+
+// k8sSchedulerStatusReasonFn is a function used by the wasm guest to set the
+// framework.Status reason.
+func k8sSchedulerStatusReasonFn(ctx context.Context, mod wazeroapi.Module, stack []uint64) {
```

Comment:
So we do actually have a host function ingesting a status. I suppose the reason this doesn't also just go ahead and take a int64 status to match native is we expect most non-zero statuses to not have reason set by the user (zero status can be just not calling the function at all) so the performance advantage of avoiding a host call is large.

If so it _might_ be something to add to the rationale perhaps others would have a similar question.

Reply:
I'll add rationale for the reduced width of status, which is that this is for custom plugins, so the custom range you can use over the built-in status codes is only 2 million ;)